lib/spack/spack/package.py

```diff
@@ -1428,6 +1578,114 @@ def rpath_args(self):
+
+
+class EditableMakefile(PackageBase):
+    phases = ['edit', 'build', 'install']
+
+    def wdir(self):
+        return self.stage.source_path
+
+    def build_args(self):
+        return list()
+
+    def install_args(self):
+        return list()
+
+    def edit(self, spec, prefix):
+        raise NotImplementedError('\'edit\' function not implemented')
+
+    def build(self, spec, prefix):
+        args = self.build_args()
+        with working_dir(self.wdir()):
+            inspect.getmodule(self).make(*args)
+
+    def install(self, spec, prefix):
+        args = self.install_args() + ['install']
+        with working_dir(self.wdir()):
+            inspect.getmodule(self).make(*args)
+
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class AutotoolsPackage(PackageBase):
+    phases = ['autoreconf', 'configure', 'build', 'install']
+
+    def autoreconf(self, spec, prefix):
+        """Not needed usually, configure should be already there"""
+        pass
+
+    @PackageBase.sanity_check('autoreconf')
+    def is_configure_or_die(self):
+        if not os.path.exists('configure'):
+            raise RuntimeError('configure script not found in {0}'.format(os.getcwd()))
+
+    def configure_args(self):
+        return list()
+
+    def configure(self, spec, prefix):
+        options = ['--prefix={0}'.format(prefix)] + self.configure_args()
+        inspect.getmodule(self).configure(*options)
+
+    def build(self, spec, prefix):
+        inspect.getmodule(self).make()
+
+    def install(self, spec, prefix):
+        inspect.getmodule(self).make('install')
+
+    # This will be used as a registration decorator in user
+    # packages, if need be
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class CMakePackage(PackageBase):
+    phases = ['cmake', 'build', 'install']
+
+    def build_type(self):
+        return 'RelWithDebInfo'
+
+    @property
+    def std_cmake_args(self):
+        # standard CMake arguments
+        args = ['-DCMAKE_INSTALL_PREFIX:PATH={0}'.format(self.prefix),
+                '-DCMAKE_BUILD_TYPE:STRING={0}'.format(self.build_type())]
+        if platform.mac_ver()[0]:
+            args.append('-DCMAKE_FIND_FRAMEWORK:STRING=LAST')
+
+        # Set up CMake rpath
+        args.append('-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=FALSE')
+        rpaths = ':'.join(spack.build_environment.get_rpaths(self))
+        args.append('-DCMAKE_INSTALL_RPATH:STRING={0}'.format(rpaths))
+        return args
+
+    def wdir(self):
+        return join_path(self.stage.source_path, 'spack-build')
+
+    def cmake_args(self):
+        return list()
+
+    def cmake(self, spec, prefix):
+        options = [self.source_directory] + self.std_cmake_args + self.cmake_args()
+        create = not os.path.exists(self.wdir())
+        with working_dir(self.wdir(), create=create):
+            inspect.getmodule(self).cmake(*options)
```

Comment:
> It is assumed to be in self.source_directory, the first argument of options

how abut adding `cmakelistdir` (or alike) with default implementation of `self.source_directory`? Just in case there will be fancy cmake packages.